src/hera/workflows/runner.py

```diff
@@ -0,0 +1,127 @@
+import argparse
+import functools
+import importlib
+import inspect
+import json
+from pathlib import Path
+from typing import Any
+
+from pydantic import validate_arguments
+
+from hera.shared.serialization import serialize
+
+
+def _ignore_unmatched_kwargs(f):
+    """Make function ignore unmatched kwargs.
+
+    If the function already has the catch all **kwargs, do nothing.
+    """
+    if _contains_var_kwarg(f):
+        return f
+
+    @functools.wraps(f)
+    def inner(**kwargs):
+        # filter out kwargs that are not part of the function signature
+        # and transform them to the correct type
+        filtered_kwargs = {key: _parse(value, key, f) for key, value in kwargs.items() if _is_kwarg_of(key, f)}
+        return f(**filtered_kwargs)
+
+    return inner
+
+
+def _contains_var_kwarg(f):
+    return any(param.kind == inspect.Parameter.VAR_KEYWORD for param in inspect.signature(f).parameters.values())
+
+
+def _is_kwarg_of(key, f):
+    param = inspect.signature(f).parameters.get(key)
+    return param and (
+        param.kind is inspect.Parameter.KEYWORD_ONLY or param.kind is inspect.Parameter.POSITIONAL_OR_KEYWORD
+    )
+
+
+def _parse(value, key, f):
+    """Parse a value to the correct type.
+
+    Args:
+        value: The value to parse.
+        key: The name of the kwarg.
+        f: The function to parse the value for.
+
+    Returns:
+        The parsed value.
+
+    """
+    if _is_str_kwarg_of(key, f):
+        return value
+    try:
+        return json.loads(value)
+    except json.JSONDecodeError:
+        return value
```

Comment:
We do and additional json.loads on all input params since argo can only contain string values.

If pydantic is passed a json encoded string, it will not coerce it. So we need to convert it to a dictionary which then lets pydantic coerce it into the right type.

Reply:
We do not json parse the type if the input argument is explicitly defined as string.